trp_project_scrum/controllers/main.py

```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import http
+from odoo.http import request
+from odoo.addons.website.controllers import main
+
+
+class Website(main.Website):
+
+    @http.route(website=True, auth="public")
```

Comment:
So basically I did this because of what happened on https://github.com/OCA/server-tools/issues/1532#issuecomment-477679348

Looking at the documentation about the decorators to validate, I see that basically they are evaluated at compile time, they take the function they decorate as an argument, and their return value is basically a callable that gets called instead of the original function. So by going through http.py, the documentation and ormcache I understand the following (which probably needs to be validated as well from @hbrunn )

1) If we override a function which has a decorator, then there is no need for us to redefine the decorator in any way (I have validated this by looking and the stack trace without a decorator and with one on this specific case. The stack trace is the same) So removing this decorator fully would be the way to go (and indeed it works)

That is because (from my understanding) first the function and its override are applied and then the decorator. Also, I derive this from 

> the returned value is bound to the function name instead of the function object
taken from here https://docs.python.org/2.7/reference/compound_stmts.html#function-definitions

Reply:
@gfcapalbo Ok scratch that, it indeed does not work without the override. I under the impression that an update of the module was not needed. In any case I will do it as you say, but still my question remains

Reply:
> in general, decorators are never inherited in python

```
class A():           
                     
    def deco(self):  
        print 'deco' 
                     
    @deco            
    def func(self):  
        print 'func' 
                     
class B(A):          
                     
    def func(self):  
        print 'func1'
                     
B().func()           

```
This prints:
deco
func1

Doesn't this tell us that first the override is applied and then the decorator on the final result?

> technically, the result of the decoration is the function you override
Reading from the documentation:

> Decorator expressions are evaluated when the function is defined, in the scope that contains the function definition. The result must be a callable, which is invoked with the function object as the only argument. The returned value is bound to the function name instead of the function object

So here we see that the https://github.com/OCA/OCB/blob/10.0/odoo/http.py#L526 does *not* in fact return the function that is being decorated rather a wrapper of it. So maybe this is the problem (not really problem by why it does not work as I expect it)?

Reply:
start time =  python interpreter reads the code, program starts.  decorator is called, so your print is called at start time. not runtime. 
the example is confusing because a decorator that returns a print isn't useful, so it's misleading.
return a function and you will reason better on this.